### PR TITLE
feat: use recurring L0 end time as hint for split times

### PR DESCRIPTION
**Pull request by Joe-Blount**
Its possible after we do vertical splitting, another L0 is written, causing the split times to shift.  Or if we didn't complete the vertical splitting, we might compute different splint times when we resume.  Either case can lead to higher than necessary write amplification as we resplit the previously split files.

This PR notices recurring L0 file end times and uses them as hints for selecting split times.

Commit 1 adds a test case simulating 11 max sized L0s being split, then another L0 arriving to shift the split points.
Commit 2 adds detection of recurring L0 end times to use as split hints, so we resplit at the same times as previous
Commit 3 shows the insta test updates.

The test added in the first commit to capture the target situation shows significant improvement in write amplification (2.53 GB  -> 1.3 GB) because it eliminates the re-splitting of the previously split L0s.  Several other tests are slightly affected (increasing or decreasing write amplification by a tiny amount).

Helps https://github.com/influxdata/idpe/issues/17974
May also help https://github.com/influxdata/EAR/issues/4468

- [x] I've read the contributing section of the project [README](https://github.com/influxdata/influxdb_iox/blob/main/README.md).
- [x] Signed [CLA](https://influxdata.com/community/cla/) (if not already signed).

### CODE REVIEW

File: compactor/tests/layouts/large_files.rs [hunk 1]
```diff
@@ -1575,488 +1575,22 @@ async fn pre_split_large_l0() {
...
-    - "**** Simulation run 65, type=split(VerticalSplit)(split_times=[9161]). 1 Input Files, 14mb total:"
-    - "L0, all files 14mb                                                                                                 "
-    - "L0.77[8705,10000] 10.01us|-----------------------------------------L0.77------------------------------------------|"
-    - "**** 2 Output Files (parquet_file_id not yet assigned), 14mb total:"
-    - "L0                                                                                                                 "
-    - "L0.?[8705,9161] 10.01us 5mb|------------L0.?-------------|                                                           "
-    - "L0.?[9162,10000] 10.01us 9mb                               |--------------------------L0.?--------------------------| "
-    - "**** Simulation run 66, type=split(VerticalSplit)(split_times=[3431, 4577, 5723, 6869, 8015, 9161, 10312]). 1 Input Files, 100mb total:"
+    - "**** Simulation run 0, type=split(VerticalSplit)(split_times=[3568, 4852, 6136, 7420, 8704, 10312]). 1 Input Files, 100mb total:"
```
Comment (wiedld) [inline compactor/tests/layouts/large_files.rs:1578]: Nice. The split times now match the file edges.

Comment (wiedld) [inline compactor/tests/layouts/large_files.rs:1578]: Nice. The split times now match the file edges.

File: compactor/tests/layouts/large_files.rs [hunk 2]
```diff
@@ -2076,264 +1610,172 @@ async fn pre_split_large_l0() {
...
+    - "L2.?[10313,12462] 11us 21mb|--------------------------------L2.?---------------------------------|                   "
+    - "L2.?[12463,13000] 11us 5mb                                                                        |-----L2.?------| "
     - "Committing partition 1:"
-    - "  Soft Deleting 3 files: L1.234, L1.235, L1.236"
-    - "  Upgrading 8 files level to CompactionLevel::L2: L1.226, L1.227, L1.228, L1.229, L1.230, L1.231, L1.232, L1.233"
+    - "  Soft Deleting 2 files: L1.100, L1.101"
+    - "  Upgrading 14 files level to CompactionLevel::L2: L1.86, L1.87, L1.88, L1.89, L1.90, L1.91, L1.92, L1.93, L1.94, L1.95, L1.96, L1.97, L1.98, L1.99"
     - "  Creating 2 files"
-    - "**** Final Output Files (2.53gb written)"
+    - "**** Final Output Files (1.3gb written)"
```
Comment (wiedld) [inline compactor/tests/layouts/large_files.rs:1761]: Nice. A lot fewer rounds and less write amplification. 🎉

File: compactor/tests/layouts/backfill.rs [hunk 3]
```diff
@@ -721,47 +721,47 @@ async fn random_backfill_empty_partition() {
...
-    - "L2.?[920,972] 1.05us 21mb|--------------------------------L2.?--------------------------------|                    "
-    - "L2.?[973,986] 1.05us 6mb                                                                         |-----L2.?------| "
+    - "L2.?[878,964] 1.05us 35mb|--------------------------------L2.?---------------------------------|                   "
+    - "L2.?[965,986] 1.05us 9mb                                                                         |-----L2.?------| "
     - "Committing partition 1:"
     - "  Soft Deleting 1 files: L1.196"
     - "  Upgrading 5 files level to CompactionLevel::L2: L1.195, L1.197, L1.198, L1.199, L1.200"
     - "  Creating 2 files"
-    - "**** Final Output Files (1.37gb written)"
+    - "**** Final Output Files (1.39gb written)"
```
Comment (wiedld) [inline compactor/tests/layouts/backfill.rs:756]: Performs about the same on random backfill. 🆗

**Review by wiedld — APPROVED** [review]
Thanks for breaking this up into commits, and demonstrating the change in the added test case.

Can we consider adding additional testing beyond the layout tests? That way we can separately demonstrate both the specific contract (e.g. how split times are determined), versus the downstream effects?

File: compactor/tests/layouts/large_files.rs [hunk 4]
```diff
@@ -1383,3 +1383,957 @@ async fn tiny_time_range() {
     "###
     );
 }
+
+// This test simulates a situation where we vertically split 11 large L0s, into 7 files each, then another L0 file is written
+// which shifts the default split points.  The purpose of this test is to see if the compactor will adjust its split points
+// to the previusly used times.  Otherwise, write amplification will be significantly impacted.
+#[tokio::test]
+async fn pre_split_large_l0() {
```
[the remark this answers is not in the document]
Comment (Joe-Blount) [reply compactor/tests/layouts/large_files.rs:1391]: I added a feature to the simulator to verify specific split times occur.  It was easy to add and is very reusable - now we can add `.with_required_split_times(vec![ ])` to any simulator test to confirm expected split times are used.

**Pull request merged**